### PR TITLE
SW_601 Include shop system and version in the Findologic request 

### PR DESCRIPTION
**Pull request by D3xter7**

### CODE REVIEW

File: tests/Findologic/Request/NavigationRequestFactoryTest.php [hunk 1]
```diff
@@ -25,13 +26,15 @@ class NavigationRequestFactoryTest extends TestCase
...
     {
+        $_SERVER['HTTP_CLIENT_IP'] = '192.168.0.1';
         $cacheKey = 'finsearch_version';
         $expectedReferer = 'http://localhost.shopware';
         $expectedIpAddress = '192.168.0.1';
         $expectedHost = 'findologic.de';
         $expectedAdapter = 'XML_2.1';
         $expectedVersion = '0.1.0';
         $expectedCategoryPath = 'Kids & Music_Computers & Shoes';
+        $expectedShopType = "Shopware6";
```
Comment (zaifastafa) [inline tests/Findologic/Request/NavigationRequestFactoryTest.php:37]: Use single quotes.

File: tests/Findologic/Request/NavigationRequestFactoryTest.php [hunk 2]
```diff
@@ -25,13 +26,15 @@ class NavigationRequestFactoryTest extends TestCase
      */
     public function testNavigationRequestInstance(): void
     {
+        $_SERVER['HTTP_CLIENT_IP'] = '192.168.0.1';
```
Comment (zaifastafa) [inline tests/Findologic/Request/NavigationRequestFactoryTest.php:29]: Why was this added? How was it working before? Because this is an existing test, and we are also checking for expected IP address, so why was this needed to be added?

File: tests/Findologic/Request/NavigationRequestFactoryTest.php [hunk 3]
```diff
@@ -71,6 +74,8 @@ public function testNavigationRequestInstance(): void
         $this->assertSame($expectedIpAddress, $params['userip']);
         $this->assertSame($expectedAdapter, $params['outputAdapter']);
         $this->assertSame($expectedHost, $params['shopurl']);
+        $this->assertSame($expectedShopType, $params['shopType']);
+        $this->assertSame(Utils::getCleanShopwareVersion(), $params['shopVersion']);
```
Comment (zaifastafa) [inline tests/Findologic/Request/NavigationRequestFactoryTest.php:78]: For consistency with other expected variables, add the Shopware version also in a variable above (similar to `$expectedShopType` and use that here. This is also fine, but its better to be consistent if making changes in an existing test so we can have all our expected variables in one place.

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 4]
```diff
@@ -74,6 +74,9 @@ protected function setDefaults(
         $searchNavigationRequest->setUserIp($this->fetchClientIp());
         $searchNavigationRequest->setRevision($this->getPluginVersion());
         $searchNavigationRequest->setOutputAdapter(OutputAdapter::XML_21);
+        $searchNavigationRequest->addIndividualParam("shopType", "Shopware6", true);
+        $searchNavigationRequest->addIndividualParam("shopVersion", Utils::getCleanShopwareVersion(), true);
```
Comment (zaifastafa) [inline src/Findologic/Request/FindologicRequestFactory.php:78]: Use single quotes.

**Review by zaifastafa — CHANGES_REQUESTED** [review]
@D3xter7 Overall much better than the previous PR 👍🏼 Please delete that obsolete one.

File: tests/Findologic/Request/NavigationRequestFactoryTest.php (hunk 2 above)
Comment (D3xter7) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php:29]: While executing there is an error showing 'invalid parameter userip' so I checked that this array was not set, that's why I assigned it manually.

Comment (zaifastafa) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php:29]: Yes I understand that, but why was this change needed in this story? It does not seem relevant to the story and it is an existing test so it should be passing already without this change.

Comment (D3xter7) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php:29]: I'll have look on it once again, I know it shouldn't give an error but it was like that,

File: src/Findologic/Request/FindologicRequestFactory.php (hunk 4 above)
Comment (zaifastafa) [reply src/Findologic/Request/FindologicRequestFactory.php]: @D3xter7 this is not resolved.

Comment (zaifastafa) [reply src/Findologic/Request/FindologicRequestFactory.php]: What does the 3rd parameter do? Why have you passed `true` in here?

File: tests/Findologic/Request/NavigationRequestFactoryTest.php (hunk 2 above)
Comment (D3xter7) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php:29]: error
While I am running a test, I am having this error after removing this line

File: src/Findologic/Request/FindologicRequestFactory.php (hunk 4 above)
Comment (D3xter7) [reply src/Findologic/Request/FindologicRequestFactory.php]: > 

I have done this in the second last commit.

Comment (D3xter7) [reply src/Findologic/Request/FindologicRequestFactory.php]: > 

I didn't test that why it is required, but this function is working fine and required the third parameter as boolean.

File: tests/Findologic/Request/NavigationRequestFactoryTest.php (hunk 2 above)
Comment (D3xter7) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php]: I have removed this line because this was working fine on the git test but not locally.

File: src/Findologic/Request/FindologicRequestFactory.php (hunk 4 above)
Comment (zaifastafa) [reply src/Findologic/Request/FindologicRequestFactory.php]: @D3xter7 the double quote is not fixed, please check again. When a comment as such is made, please make sure to check other places too where this is applicable as previously also mentioned to you by Dominik.

**Review by zaifastafa — APPROVED** [review]

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 5]
```diff
@@ -74,6 +74,9 @@ protected function setDefaults(
         $searchNavigationRequest->setUserIp($this->fetchClientIp());
         $searchNavigationRequest->setRevision($this->getPluginVersion());
         $searchNavigationRequest->setOutputAdapter(OutputAdapter::XML_21);
+        $searchNavigationRequest->addIndividualParam('shopType', 'Shopware6', true);
+        $searchNavigationRequest->addIndividualParam('shopVersion', Utils::getCleanShopwareVersion(), true);
```
Comment (TheKeymaster) [inline src/Findologic/Request/FindologicRequestFactory.php:78]: How can this work? The third parameter of this method is `$method` and it expects a `string`.

image

File: tests/Findologic/Request/NavigationRequestFactoryTest.php [hunk 6]
```diff
@@ -71,6 +74,8 @@ public function testNavigationRequestInstance(): void
         $this->assertSame($expectedIpAddress, $params['userip']);
         $this->assertSame($expectedAdapter, $params['outputAdapter']);
         $this->assertSame($expectedHost, $params['shopurl']);
+        $this->assertSame($expectedShopType, $params['shopType']);
+        $this->assertSame($expectedShopVersion, $params['shopVersion']);
```
Comment (TheKeymaster) [inline tests/Findologic/Request/NavigationRequestFactoryTest.php:78]: Why are these asserted before the `cat`? To have the same structure at the top as at the assertions down here, either move these two assertions after the `cat` assertion, or move the declarations at the top above the `$expectedCategoryPath` declaration.

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: tests/Findologic/Request/NavigationRequestFactoryTest.php (hunk 6 above)
Comment (D3xter7) [reply tests/Findologic/Request/NavigationRequestFactoryTest.php:79]: What I have done is, I have declared '$expectedShopType' and 'expectedShopVersion' variables after '$expectedCategoryPath', and the same structure goes for assertions

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 7]
```diff
@@ -74,6 +74,9 @@ protected function setDefaults(
         $searchNavigationRequest->setUserIp($this->fetchClientIp());
         $searchNavigationRequest->setRevision($this->getPluginVersion());
         $searchNavigationRequest->setOutputAdapter(OutputAdapter::XML_21);
+        $searchNavigationRequest->addIndividualParam('shopType', 'Shopware6', true);
```
Comment (zaifastafa) [inline src/Findologic/Request/FindologicRequestFactory.php:77]: The third parameter here is not fixed in this line.

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 8]
```diff
@@ -74,6 +74,9 @@ protected function setDefaults(
...
+        $searchNavigationRequest->addIndividualParam('shopVersion', Utils::getCleanShopwareVersion(), 'set');
```
Comment (dblaichinger) [inline src/Findologic/Request/FindologicRequestFactory.php:78]: In the docblock of `addIndividualParam` it clearly states:

>   * @param $method string Use Request::ADD_VALUE to add the param and not overwrite existing ones and
      Request::SET_VALUE to overwrite existing params.

Why using a hardcoded string when have the constant `Request::SET_VALUE`?

**Review by dblaichinger — CHANGES_REQUESTED** [review]

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 9]
```diff
@@ -6,6 +6,7 @@
 
 use FINDOLOGIC\Api\Definitions\OutputAdapter;
 use FINDOLOGIC\Api\Exceptions\InvalidParamException;
+use FINDOLOGIC\Api\Requests\Request as FindoLogicApiRequest;
```
Comment (TheKeymaster) [inline src/Findologic/Request/FindologicRequestFactory.php:9]: It is not `FindoLogic`, but `Findologic`
```suggestion
use FINDOLOGIC\Api\Requests\Request as FindologicApiRequest;
```

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 10]
```diff
@@ -74,6 +50,13 @@ protected function setDefaults(
         $searchNavigationRequest->setUserIp($this->fetchClientIp());
         $searchNavigationRequest->setRevision($this->getPluginVersion());
         $searchNavigationRequest->setOutputAdapter(OutputAdapter::XML_21);
+        $searchNavigationRequest->addIndividualParam('shopType', 'Shopware6', FindoLogicApiRequest::ADD_VALUE);
+        $searchNavigationRequest->addIndividualParam(
+            'shopVersion',
+            Utils::getCleanShopwareVersion(),
+            FindoLogicApiRequest::ADD_VALUE
+        );
```
Comment (TheKeymaster) [inline src/Findologic/Request/FindologicRequestFactory.php:58]: This should not be `ADD_VALUE` but `SET_VALUE`, as this parameter should only be set once and not be able to have multiple values.
```suggestion
        $searchNavigationRequest->addIndividualParam('shopType', 'Shopware6', FindoLogicApiRequest::SET_VALUE);
        $searchNavigationRequest->addIndividualParam(
            'shopVersion',
            Utils::getCleanShopwareVersion(),
            FindoLogicApiRequest::SET_VALUE
        );
```

File: src/Findologic/Request/FindologicRequestFactory.php [hunk 11]
```diff
@@ -127,6 +110,31 @@ private function fetchClientIp(): string
         return $ipAddress;
     }
 
+    /**
+     * @throws InvalidArgumentException
+     * @throws InconsistentCriteriaIdsException
+     */
+    protected function getPluginVersion(): string
```
Comment (TheKeymaster) [inline src/Findologic/Request/FindologicRequestFactory.php:117]: Why was this method moved here? If your intention was to move the `setDefaults` to the top, then please move this method above `fetchClientIp`, as it is a private method.

**Review by TheKeymaster — CHANGES_REQUESTED** [review]

File: src/Findologic/Request/FindologicRequestFactory.php (hunk 11 above)
Comment (D3xter7) [reply src/Findologic/Request/FindologicRequestFactory.php:117]: I had no intentions to move it anywhere, now it is above the `fetchClientIp` method

Comment (TheKeymaster) [reply src/Findologic/Request/FindologicRequestFactory.php:135]: This wasn't resolved.

Comment (D3xter7) [reply src/Findologic/Request/FindologicRequestFactory.php:135]: @TheKeymaster Do I have to make them in sequence? like, all protected functions then all private functions?

Comment (TheKeymaster) [reply src/Findologic/Request/FindologicRequestFactory.php:135]: @D3xter7 either that or you revert the change, if you didn't intend to change it after all.

Comment (D3xter7) [reply src/Findologic/Request/FindologicRequestFactory.php]: @TheKeymaster I have made their sequence, protected -> private.

**Review by TheKeymaster — APPROVED** [review]

**Review by dblaichinger — APPROVED** [review]
